README.md
<a href="https://www.taoensso.com" title="More stuff by @ptaoussanis at www.taoensso.com">
<img src="https://www.taoensso.com/taoensso-open-source.png" alt="Taoensso open-source" width="400"/></a>

**[CHANGELOG]** | [API] | current [Break Version]:

```clojure
[com.taoensso/tower "3.0.2"]       ; Deprecated
[com.taoensso/tower "3.1.0-beta4"] ; EOL but stable
```

> See [here](https://taoensso.com/clojure/backers) if you're interested in helping support my open-source work, thanks! - Peter Taoussanis

# Tower: a Clojure/Script i18n & L10n library

The Java platform provides some very capable tools for writing internationalized applications. Unfortunately, they can be... cumbersome. We can do much better in Clojure.

Tower's an attempt to present a **simple, idiomatic internationalization and localization** story for Clojure. It wraps standard Java functionality where possible - with warm, fuzzy, functional love.

## Library status: EOL but stable

**Last updated:** Jan 2016

Tower's latest beta (`3.1.0-beta4`) is **stable and quite useable in production**, but is likely the last major release for the library (modulo unexpected bug fixes). **Future development work** is going to be focused on:

  * [Tempura] - Clojure/Script **translations** API
  * To be determined - Clojure/Script **localization** API

##### Why the EOL?

Tower's quite useable as it is but was initially designed before things like Reactjs (indeed ClojureScript) were available. The API does the job, but it's not what I'd write if I wrote it from scratch today w/o the historical baggage.

Rather than try to shoehorn a significant set of potentially-breaking changes into Tower (pressuring stable users to migrate), it seemed logical to instead focus future work on alternative lib/s that could be **optionally migrated to in stages**.

As I find time to publish these libs, I'll make a best effort to also document the procedure necessary for **migrating from Tower** for those users that want to do that.

I apologise for the stress/trouble that this might cause, but hope that the present solution will turn out to be relatively painless in most cases. Long term, I think that the benefits will be worth it.

\- [Peter Taoussanis]

## Features
 * Small **all-Clojure** library
 * Simple wrappers for standard Java **localization features**
 * **Simple, map-based** translation dictionary format. No XML or resource files!
 * Rails-like, pure-Clojure **translation function** with ClojureScript support
 * Automatic dev-mode **dictionary reloading** for rapid REPL development
 * Seamless **markdown support** for translators
 * **Ring middleware**
 * TODO: export/import to allow use with **industry-standard tools for translators**

## Getting started

Add the necessary dependency to your project:

```clojure
Leiningen: [com.taoensso/tower "3.0.2"] ; or
deps.edn:   com.taoensso/tower {:mvn/version "3.0.2"}
```

And setup your namespace imports:

```clojure
(ns my-clj-ns ; Clojure namespace
  (:require [taoensso.tower :as tower :refer (with-tscope)]))

;; Requires v3.1+
(ns my-cljs-ns ; ClojureScript namespace
  (:require [taoensso.tower :as tower :refer-macros (with-tscope)]))
```

### Translation

The `make-t` fn handles translations. You give it a config map which includes your dictionary, and get back a `(fn [locale-or-locales k-or-ks & fmt-args])`:

```clojure
(def my-tconfig
  {:dictionary ; Map or named resource containing map
   {:en   {:example {:foo         ":en :example/foo text"
                     :foo_comment "Hello translator, please do x"
                     :bar {:baz ":en :example.bar/baz text"}
                     :greeting "Hello %s, how are you?"
                     :inline-markdown "<tag>**strong**</tag>"
                     :block-markdown* "<tag>**strong**</tag>"
                     :with-exclaim!   "<tag>**strong**</tag>"
                     :with-arguments  "Num %d = %s"
                     :greeting-alias :example/greeting
                     :baz-alias      :example.bar/baz}
           :missing  "|Missing translation: [%1$s %2$s %3$s]|"}
    :en-US {:example {:foo ":en-US :example/foo text"}}
    :de    {:example {:foo ":de :example/foo text"}}
    :ja "test_ja.clj" ; Import locale's map from external resource
    }
   :dev-mode? true ; Set to true for auto dictionary reloading
   :fallback-locale :de})

(def t (tower/make-t my-tconfig)) ; Create translation fn

(t :en-US :example/foo) => ":en-US :example/foo text"
(t :en    :example/foo) => ":en :example/foo text"
(t :en    :example/greeting "Steve") => "Hello Steve, how are you?"

;;; Translation strings are escaped and parsed as inline or block Markdown:
(t :en :example/inline-markdown) => "&lt;tag&gt;<strong>strong</strong>&lt;/tag&gt;"
(t :en :example/block-markdown)  => "<p>&lt;tag&gt;<strong>strong</strong>&lt;/tag&gt;</p>" ; Notice no "*" suffix here, only in dictionary map
(t :en :example/with-exclaim)    => "<tag>**strong**</tag>" ; Notice no "!" suffix here, only in dictionary map
(t :en :example/with-arguments 42 "forty two") =>   "Num 42 = forty two"
```

It's simple to get started, but there's a number of advanced features for if/when you need them:

#### Loading dictionaries from disk/resources

Just use a string for the `:dictionary` and/or any locale value(s) in your config map. Be sure to check that the appropriate files are available on your classpath or one of Leiningen's resource paths (e.g. `resources/`).

#### Reloading dictionaries on modification**

Enable the `:dev-mode?` option and you're good to go!

#### Scoping translations**

Use `with-tscope` if you're calling `t` repeatedly within a specific translation-namespace context:
```clojure
(with-tscope :example
  [(t :en :foo)
   (t :en :bar/baz)]) => [":en :example/foo text" ":en :example.bar/baz text"]
```

#### Missing translations**

These are handled gracefully. `(t :en-US :example/foo)` will search for a translation as follows:

 1. `:example/foo` in the `:en-US` locale.
 2. `:example/foo` in the `:en` locale.
 3. `:example/foo` in the dictionary's fallback locale.
 4. `:missing` in any of the above locales.

You can also specify fallback keys that'll be tried before other locales. `(t :en-US [:example/foo :example/bar]))` searches:

 1. `:example/foo` in the `:en-US` locale.
 2. `:example/bar` in the `:en-US` locale.
 3. `:example/foo` in the `:en` locale.
 4. `:example/bar` in the `:en` locale.
 5. `:example/foo` in the fallback locale.
 6. `:example/bar` in the fallback locale.
 7. `:missing` in any of the above locales.

And even fallback locales. `(t [:fr-FR :en-US] :example/foo)` searches:

 1. `:example/foo` in the `:fr-FR` locale.
 2. `:example/foo` in the `:fr` locale.
 3. `:example/foo` in the `:en-US` locale.
 4. `:example/foo` in the `:en` locale.
 5. `:example/foo` in the fallback locale.
 6. `:missing` in any of the above locales.

In all cases, translation requests are logged upon fallback to fallback locale or :missing key.

#### ClojureScript translations (early support, v3.1+)

```clojure
(def ^:private tconfig
  {:fallback-locale :en
   ;; Inlined (macro) dict => this ns needs rebuild for dict changes to reflect.
   ;; (dictionary .clj file can be placed in project's `/resources` dir):
   :compiled-dictionary (tower-macros/dict-compile* "my-dict.clj")})

(def t (tower/make-t tconfig)) ; Create translation fn

(t :en-US :example/foo) => ":en-US :example/foo text"
```

There's two notable differences from JVM translations:

 1. The dictionary is provided in a _pre-compiled_ form so that it can be inlined directly into your Cljs.
 2. Since we lack a locale-aware Cljs `format` fn, your translations _cannot_ use JVM locale formatting patterns.

The API is otherwise exactly the same, including support for all decorators.

##### Use with React (Reagent/Om/etc.)

React presents a bit of a challenge to translations since it **automatically escapes all text content** as a security measure.

This has two important implications for use with Tower's translations:

 1. Content intended to allow _translator-controlled inline styles_ needs to provided to React with the `dangerouslySetInnerHTML` property.
 2. All other content should get a `:<key>!`-style translation to prevent double escaping (Tower already escapes translations not marked with an exlamation point).

### Localization

Check out `fmt`, `parse`, `lsort`, `fmt-str`, `fmt-msg`:

```clojure
(tower/fmt   :en-ZA 200       :currency) => "R 200.00"
(tower/fmt   :en-US 200       :currency) => "$200.00"
(tower/parse :en-US "$200.00" :currency) => 200

(tower/fmt :de-DE 2000.1 :number)               => "2.000,1"
(tower/fmt :de-DE (java.util.Date.))            => "12.06.2012"
(tower/fmt :de-DE (java.util.Date.) :date-long) => "12. Juni 2012"
(tower/fmt :de-DE (java.util.Date.) :dt-long)   => "12 giugno 2012 16.48.01 ICT"

(tower/lsort :pl ["Warsaw" "Kraków" "Łódź" "Wrocław" "Poznań"])
=> ("Kraków" "Łódź" "Poznań" "Warsaw" "Wrocław")

(mapv #(tower/fmt-msg :de "{0,choice,0#no cats|1#one cat|1<{0,number} cats}" %)
        (range 5))
=> ["no cats" "one cat" "2 cats" "3 cats" "4 cats"]
```

It's that simple. See the [API] docs for details.

#### Country and languages names, timezones, etc.

Check out `countries`, `languages`, and `timezones`.

### Ring middleware

Quickly internationalize your Ring web apps by adding `taoensso.tower.ring/wrap-tower` to your middleware stack.

See the [API] docs for details.

## This project supports the ![ClojureWerkz-logo] goals

 * [ClojureWerkz] is a growing collection of open-source, **batteries-included Clojure libraries** that emphasise modern targets, great documentation, and thorough testing.

## Contacting me / contributions

Please use the project's [GitHub issues page] for all questions, ideas, etc. **Pull requests welcome**. See the project's [GitHub contributors page] for a list of contributors.

Otherwise, you can reach me at [Taoensso.com]. Happy hacking!

\- [Peter Taoussanis]

## License

Distributed under the [EPL v1.0] \(same as Clojure).  
Copyright &copy; 2012-2020 [Peter Taoussanis].

<!--- Standard links -->
[Taoensso.com]: https://www.taoensso.com
[Peter Taoussanis]: https://www.taoensso.com
[@ptaoussanis]: https://www.taoensso.com
[More by @ptaoussanis]: https://www.taoensso.com
[Break Version]: https://github.com/taoensso/encore/blob/master/BREAK-VERSIONING.md

<!--- Standard links (repo specific) -->
[CHANGELOG]: https://github.com/taoensso/tower/releases
[API]: http://taoensso.github.io/tower/
[GitHub issues page]: https://github.com/taoensso/tower/issues
[GitHub contributors page]: https://github.com/taoensso/tower/graphs/contributors
[EPL v1.0]: https://raw.githubusercontent.com/ptaoussanis/tower/master/LICENSE
[Hero]: https://raw.githubusercontent.com/ptaoussanis/tower/master/hero.png "Title"

<!--- Unique links -->
[Tempura]: https://github.com/taoensso/tempura
[ClojureWerkz-logo]: https://raw.github.com/clojurewerkz/clojurewerkz.org/master/assets/images/logos/clojurewerkz_long_h_50.png
[ClojureWerkz]: http://clojurewerkz.org/


CHANGELOG.md
This project uses Break Versioning (https://www.taoensso.com/break-versioning)

## v3.1.0-beta4 / 2015 Jul 13

> As beta3 with updated deps

```clojure
[com.taoensso/tower "3.1.0-beta4"]
```


## v3.1.0-beta3 / 2015 Apr 28

> Non-breaking hotfix release

* **Fix**: broken support for invalid JVM locales in translation dicts [@okal #62]

```clojure
[com.taoensso/tower "3.1.0-beta3"]
```


## v3.1.0-beta2 / 2015 Mar 3

 > Non-breaking hotfix release

 * **Fix**: unnested resource dictionaries were not loading correctly [@juhani-hietikko #60]

```clojure
[com.taoensso/tower "3.1.0-beta2"]
```

## v3.1.0-beta1 / 2015 Feb 28

 > This is a **major update** that **may be BREAKING**.

 * **BREAK**: ClojureScript dict compiler has been renamed `dict-compile`->`dict-compile*` for consistency with new `dict-load`,`dict-load*` fn+macro pair.
 * **BREAK** (rarely): `t` now _always_ returns a string unless it has an explicit `nil` fallback.
 * **BREAK** (rarely): Fix strange JVM handling for locales #{:he :yi :id}. [@juhani-hietikko - #56]
 * **New**: add `get-countries`, `get-langs` (**experimental**).
 * **New**: add `kw-locale` `:lang-only?` option.
 * **New**: add fully-configurable dictionary decorators (**experimental**).
 * **New**: eval macro-time dict compilation to allow for dicts in vars, etc. [#54]
 * **New**: `country-name`, `lang-name` are now public (**experimental**).
 * **Fix**: ensure that `foo.bar` and `foo/bar` reach same translation.
 * **Implementation**: move to unified .cljx codebase.
 * **Implementation**: `t`, Ring middleware performance improvements.


## v3.0.2 / 2014 Oct 1

 * **CHANGE**: no longer throw NPEs on `nil` format patterns (e.g. for missing translation key).


## v3.0.1 / 2014 Sep 7

 * **FIX**: https://github.com/taoensso/timbre/issues/79.


## v3.0.0 / 2014 Aug 28

> This is a **major update** that **may be BREAKING** for users upgrading from < `v2.1.0-RC1`.
> It introduces ClojureScript translation support and fixes a number of useability sharp edges.

 * **FIX**: All localization formatters are now correctly thread safe.

 * **NEW**: Added ClojureScript translation support. See the README for an example and notes.
 * **NEW**: `timezones` fn now supports optional timezone-ids arg.
 * **NEW**: Add `all-timezone-ids` set.
 * **NEW** [#42]: Translation dictionary now supports underscores in translation keys.
 * **NEW** [#43]: Translation fns can now take a _vector_ of descending-preference locales (@vvvvalvalval).
 * **NEW** [#43]: Ring middleware now automatically attaches a smarter translation fn that'll search through all of a client's sorted Accept-Language header languages when looking for a translation.
 * **NEW** [#50], [#52]: Translation dictionary now supports arbitrary (non-JVM) locales.

 * **CHANGE**: Dropped (experimental) `:scope-var` tconfig option.
 * **CHANGE**: Dropped (experimental) `:root-scope` tconfig option.
 * **CHANGE**: Default :missing translations entry now avoids <>'s (no need for html escaping).
 * **CHANGE**: `languages` now returns languages as "localized (unlocalized)" pairs rather than "unlocalized (localized)" pairs.
 * **CHANGE**: All `Exception`s are now `ExceptionInfo`s.

 * **POSSIBLY BREAKING**: `translate` and `t` are both being phased out in favor of a new `make-t` fn. The new approach is more flexible and faster. This change is _non-breaking_ **if** you use the Ring middleware; otherwise please see the README for new recommended usage examples.

 * **DEPRECATED**: `locale`->`jvm-locale`, `try-locale`->`try-jvm-locale` (only the names have changed).
 * **DEPRECATED**: `wrap-tower-middleware` -> `wrap-tower`. This is a recommended change, but it's **BREAKING** if you make it:
  ```clojure
  ;;; 1. The fn signature has changed (tconfig is now an explicit arg):
  (wrap-tower-middleware <ring-handler> {:tconfig _ <other opts>}) ; Old
  ;; vs
  (wrap-tower <ring-handler> <tconfig> {<other-opts>}) ; New

  ;;; 2. The Ring request's `:t` key has changed:
  {:locale _ :t  (fn [k-or-ks & fmt-args])} ; Old
  ;; vs
  {:locale _ :t  (fn [locale k-or-ks & fmt-args]) ; Now takes a locale
             :t' (fn [k-or-ks & fmt-args])}       ; New, behaves like old `:t`
  ```

 The new behaviour is more consistent. `t` always refers to a translation fn that takes a locale arg, and `t'` always refer to a partial translation fn that has already been provided a locale arg. **Migrate** by swapping your middleware, and using `t'` instead of `t` as your locale-less translation fn. **OR** you can give a `:legacy-t? true` opt to `wrap-tower` to keep the old behaviour.


## v2.0.2 / 2014 Jan 19

> This is a **backwards compatible bug fix release**. Recommended upgrade.

### New

 * `normalize` fn now takes optional normalization form.

### Fixes

 * Broken `fmt-fn` argument for translate fn.
 * Fallback locales should have `locale-key` called on their locale.
 * #37 Broken `t` parent fallback for empty child locales.


## v2.0.0 → v2.0.1

  Minor, non-breaking release: **2013 Nov 26**.

  * [FIX] Translations: `with-tscope` was preventing `:missing` translation fallbacks from working correctly (ystael).
  * [HK] Translations: `t` now throws a proper error message when trying to format arguments against a `nil` translation pattern.


## v2.0.0-beta3 → 2.0.0

  * **BREAKING**: extra args to `t` now get formatted with `java.util.Formatter` (`fmt-str`) rather than `MessageFormat` (`fmt-msg`). You can override this preference by specifying a `:fmt-fn` config option. `oldt` retains the old behavior by default.
  * **BREAKING**: `languages` and `countries` now return lowercase keyword codes.
  * **BREAKING**: `languages`, `countries`, and `timezones` now return sorted maps.
  * Dictionaries: can now load locales from additional resource files (mopemope).
  * Dictionaries: refactored.
  * New public vars: `iso-countries`, `iso-languages`, `major-timezone-ids`.


## v1.7.1 → 2.0.0-beta3


src/tower/core.clj
(ns tower.core
  "Simple internationalization (i18n) library for Clojure. Wraps standard Java
  functionality when possible, removing unnecessary boilerplate."
  {:author "Peter Taoussanis"}
  (:require [clojure.string  :as str]
            [clojure.java.io :as io]
            [timbre.core     :as timbre])
  (:use     [tower.utils     :as utils :only (defmem-)])
  (:import  [java.util Date Locale TimeZone]
            [java.text Collator NumberFormat DateFormat]))

;;;; Default configuration

(declare compiled-dictionary)

(def config
  "This map controls everything about the way Tower operates.

  To enable translations, :dictionary should be a map of form
  {:locale {:ns1 ... {:nsN {:key<.optional-decorator> text}}}}}

  Other important options include:
    :default-locale which controls fallback locale for `with-locale` and for
    default missing-translation function.

    :dev-mode? which controls Ring middleware's automatic dictionary reloading
    and the behaviour of default missing-translation function.

  See source code for further details."
  (atom {:dev-mode?      true
         :default-locale :en

         :dictionary-compiler-options {:escape-undecorated? true}

         ;; Canonical example dictionary used for dev/debug, unit tests,
         ;; README, etc.
         :dictionary
         {:en         {:example {:foo ":en :example/foo text"
                                 :bar ":en :example/bar text"
                                 :decorated {:foo.html "<tag>"
                                             :foo.note "Translator note"
                                             :bar.md   "**strong**"
                                             :baz      "<tag>"}}}
          :en_US      {:example {:foo ":en_US :example/foo text"}}
          :en_US_var1 {:example {:foo ":en_US_var1 :example/foo text"}}}

         :missing-translation-fn
         (fn [{:keys [key locale]}]
           (let [{:keys [dev-mode? default-locale]} @config]
             (if dev-mode?
             (str "**" key "**")
             (do (timbre/error "Missing translation" key "for" locale)
                 (get-in @compiled-dictionary [default-locale key]
                         "")))))}))

(defn set-config! [[k & ks] val] (swap! config assoc-in (cons k ks) val))

;;;; Locales (big L for the Java object)

(defn- make-Locale
  "Creates a Java Locale object with a lowercase ISO-639 language code,
  optional uppercase ISO-3166 country code, and optional vender-specific variant
  code."
  ([lang]                 (Locale. lang))
  ([lang country]         (Locale. lang country))
  ([lang country variant] (Locale. lang country variant)))

(def ^:private available-Locales (set (Locale/getAvailableLocales)))

(defn parse-Locale
  "Returns valid Locale matching given name string/keyword, or nil if no valid
  matching Locale could be found. `locale` should be of form :en, :en_US,
  :en_US_variant, or :default for config's default."
  [locale]
  (when locale
    (if (= locale :default)
      (or (parse-Locale (:default-locale @config)) (Locale/getDefault))
      (let [new-Locale (apply make-Locale (str/split (name locale) #"[_-]"))]
        (when (available-Locales new-Locale) new-Locale)))))

(comment (map parse-Locale [nil :invalid :default :en_US]))

;;;; Bindings

(def ^:dynamic *Locale*            nil)
(def ^:dynamic *translation-scope* nil)

(defmacro with-locale
  "Executes body within the context of thread-local locale binding, enabling
  use of translation and localization functions. `locale` should be of form :en,
  :en_US, :en_US_variant, or :default for config's default."
  [locale & body]
  `(binding [*Locale* (or (parse-Locale ~locale)
                          (throw (Exception. (str "Invalid locale: " ~locale))))]
     ~@body))

(defmacro with-scope
  "Executes body within the context of thread-local translation-scope binding.
  `translation-scope` should be a keyword like :example/greetings, or nil."
  [translation-scope & body]
  `(binding [*translation-scope* ~translation-scope] ~@body))

;;;; Collation, etc.

(defmem- get-collator Collator [locale] (Collator/getInstance locale))

(defn l-compare "Localized Unicode comparator."
  [x y] (.compare (get-collator *Locale*) x y))

(comment (with-locale :default (sort l-compare ["a" "d" "c" "b" "f" "_"])))

(defn normalize
  "Transforms Unicode string into W3C-recommended standard de/composition form
  allowing easier searching and sorting of strings. Normalization is considered
  good hygiene when communicating with a DB or other software."
  [s] (java.text.Normalizer/normalize s java.text.Normalizer$Form/NFC))

;;;; Localized number formatting

(defmem- f-number   NumberFormat [Loc] (NumberFormat/getNumberInstance   Loc))
(defmem- f-integer  NumberFormat [Loc] (NumberFormat/getIntegerInstance  Loc))
(defmem- f-percent  NumberFormat [Loc] (NumberFormat/getPercentInstance  Loc))
(defmem- f-currency NumberFormat [Loc] (NumberFormat/getCurrencyInstance Loc))

(defn format-number   [x] (.format (f-number   *Locale*) x))
(defn format-integer  [x] (.format (f-integer  *Locale*) x))
(defn format-percent  [x] (.format (f-percent  *Locale*) x))
(defn format-currency [x] (.format (f-currency *Locale*) x))

(defn parse-number    [s] (.parse (f-number   *Locale*) s))
(defn parse-integer   [s] (.parse (f-integer  *Locale*) s))
(defn parse-percent   [s] (.parse (f-percent  *Locale*) s))
(defn parse-currency  [s] (.parse (f-currency *Locale*) s))

(comment (with-locale :en_ZA (format-currency 200))
         (with-locale :en_ZA (parse-currency "R 200.33")))

;;;; Localized date/time formatting

(defmem- f-date DateFormat [st Loc]    (DateFormat/getDateInstance st Loc))
(defmem- f-time DateFormat [st Loc]    (DateFormat/getTimeInstance st Loc))
(defmem- f-dt   DateFormat [ds ts Loc] (DateFormat/getDateTimeInstance ds ts Loc))

(defn style
  "Returns a DateFormat time/date style constant by style key e/o
  #{:default :short :medium :long :full}."
  ([] (style :default))
  ([style]
     (case style
       :default DateFormat/DEFAULT
       :short   DateFormat/SHORT
       :medium  DateFormat/MEDIUM
       :long    DateFormat/LONG
       :full    DateFormat/FULL
       (throw (Exception. (str "Unknown style: " style))))))

(defn format-date
  ([d]       (format-date (style) d))
  ([style d] (.format (f-date style *Locale*) d)))

(defn format-time
  ([t]       (format-time (style) t))
  ([style t] (.format (f-time style *Locale*) t)))

(defn format-dt
  ([dt] (format-dt (style) (style) dt))
  ([date-style time-style dt]
     (.format (f-dt date-style time-style *Locale*) dt)))

(defn parse-date
  ([s]       (parse-date (style) s))
  ([style s] (.parse (f-date style *Locale*) s)))

(defn parse-time
  ([s]       (parse-time (style) s))
  ([style s] (.parse (f-time style *Locale*) s)))

(defn parse-dt
  ([s] (parse-dt (style) (style) s))
  ([date-style time-style s]
     (.parse (f-dt date-style time-style *Locale*) s)))

(comment (with-locale :en_ZA (format-date (Date.)))
         (with-locale :en_ZA (format-date (style :full) (Date.)))
         (with-locale :en_ZA (format-time (style :short) (Date.)))
         (with-locale :en_ZA (format-dt (style :full) (style :full) (Date.))))

;;;; Localized text formatting

(defn format-str
  "Like clojure.core/format but uses a locale."
  ^String [fmt & args]
  (String/format *Locale* fmt (to-array args)))

(defn format-msg
  "Creates a localized message formatter and parse pattern string, substituting
  given arguments as per MessageFormat spec."
  ^String [pattern & args]
  (let [formatter (java.text.MessageFormat. pattern *Locale*)]
    (.format formatter (to-array args))))

(comment
  (with-locale :de (format-msg "foobar {0}!" 102.22))
  (with-locale :de (format-msg "foobar {0,number,integer}!" 102.22))
  (with-locale :de
    (-> #(format-msg "{0,choice,0#no cats|1#one cat|1<{0,number} cats}" %)
        (map (range 5)) doall)))

;;;; Localized country & language names

(def ^:private get-sorted-localized-names
  "Returns map containing ISO codes and corresponding localized names, both
  sorted by the localized names."
  (memoize
   (fn [display-fn iso-codes display-Loc]
     (let [;; [localized-display-name code] seq sorted by localized-display-name
           sorted-pairs
           (->> (for [code iso-codes] [(display-fn code) code])
                (sort (fn [[x _] [y _]]
                        (.compare (get-collator display-Loc) x y))))]
       ;; mapv requires Clojure 1.4+
       {:sorted-names (vec (map first sorted-pairs))
        :sorted-codes (vec (map second sorted-pairs))}))))

(def ^:private all-iso-countries (seq (Locale/getISOCountries)))

(defn sorted-localized-countries
  "Returns map containing ISO country codes and corresponding localized country
  names, both sorted by the localized names."
  ([] (sorted-localized-countries all-iso-countries))
  ([iso-countries]
     (get-sorted-localized-names
      (fn [code] (.getDisplayCountry (Locale. "" code) *Locale*))
      iso-countries *Locale*)))

(comment (with-locale :pl (sorted-localized-countries ["GB" "DE" "PL"])))

(def ^:private all-iso-languages (seq (Locale/getISOLanguages)))

(defn sorted-localized-languages
  "Returns map containing ISO language codes and corresponding localized
  language names, both sorted by the localized names."
  ([] (sorted-localized-languages all-iso-languages))
  ([iso-languages]
     (get-sorted-localized-names
      (fn [code] (let [Loc (Locale. code)]
                  (str (.getDisplayLanguage Loc *Locale*)
                       ;; Also provide each name in it's OWN language
                       (when (not= Loc *Locale*)
                         (str " (" (.getDisplayLanguage Loc Loc) ")")))))
      iso-languages *Locale*)))

(comment (with-locale :pl (sorted-localized-languages ["en" "de" "pl"])))

;;;; Timezones (note: doesn't use locales)

(def ^:private major-timezone-ids
  (->> (TimeZone/getAvailableIDs)
       (filter #(re-find #"^(Africa|America|Asia|Atlantic|Australia|Europe|Indian|Pacific)/.*" %))))

(defn- timezone-display-name "(GMT +05:30) Colombo"
  [city-tz-id offset]
  (let [[region city] (str/split city-tz-id #"/")
        offset-mins   (/ offset 1000 60)]
    (str "(GMT " (if (neg? offset-mins) "-" "+")
         (format "%02d:%02d"
                 (Math/abs (int (/ offset-mins 60)))
                 (mod (int offset-mins) 60))
         ") " city)))

(comment (timezone-display-name "Asia/Bangkok" (* 90 60 1000)))

(def sorted-timezones
  "Returns map containing timezone IDs and corresponding pretty timezone names,
  both sorted by the timezone's offset. Caches result for 3 hours."
  (utils/ttl-memoize
   #=(* 3 60 60 1000) ; 3hr ttl
   (fn []
     (let [;; [timezone-display-name id] seq sorted by timezone's offset
           sorted-pairs
           (->> (for [id major-timezone-ids]
                  (let [instant (System/currentTimeMillis)
                        tz      (TimeZone/getTimeZone id)
                        offset  (.getOffset tz instant)]
                    [offset (timezone-display-name id offset) id]))
                (sort-by first)
                (map (comp vec rest)))]
       {:sorted-names (vec (map first sorted-pairs))
        :sorted-ids   (vec (map second sorted-pairs))}))))

(comment (take 10      (:sorted-names (sorted-timezones)))
         (take-last 10 (:sorted-names (sorted-timezones))))

;;;; Dictionary management

(def ^:private compiled-dictionary
  "Compiled form of (:dictionary @config) that:
    1. Is efficient to store.
    2. Is efficient to access.
    3. Has display-ready, decorator-controlled text values."
  (atom {}))

(defn load-dictionary-from-map-resource!
  "Sets dictionary by reading Clojure map from named resource. Without any
  arguments, searches for `tower-dictionary.clj` in classpath and Leiningen's
  resource paths."
  ([] (load-dictionary-from-map-resource! "tower-dictionary.clj"))
  ([resource-name]
     (try (->> resource-name
               io/resource
               io/reader
               slurp
               read-string
               (set-config! [:dictionary]))
          ;; For Ring middleware auto-reloading:
          (set-config! [:dict-res-name] resource-name))))

(defn- compile-map-path
  "[:locale :ns1 ... :nsN unscoped-key.decorator translation] =>
  {:locale {:ns1/.../nsN/unscoped-key (f translation decorator)}}"
  [{:keys [escape-undecorated?] :as compiler-options} path]
  {:pre [(seq path) (>= (count path) 3)]}
  (let [path        (vec path)
        locale-name (first path)
        translation (peek path)
        scope-ks    (subvec path 1 (- (count path) 2)) ; [:ns1 ... :nsN]

        ;; Check for possible decorator
        [unscoped-k decorator]
        (->> (str/split (name (peek (pop path))) #"[\._]")
             (map keyword))

        ;; [:ns1 ... :nsN :unscoped-key] => :ns1/.../nsN/unscoped-key
        scoped-key (->> (conj scope-ks unscoped-k)
                        (map name)
                        (str/join "/")
                        (keyword))]

    (when (not= decorator :note) ; Discard translator notes
      {locale-name
       {scoped-key
        (case decorator
          :html translation ; HTML-safe, leave unchanged

          ;; Escape and parse as inline markdown
          :md (-> (str translation)
                  (utils/escape-html)
                  (utils/inline-markdown->html))

          ;; No decorator
          (if escape-undecorated?
            (utils/escape-html translation)
            translation))}})))

(defn- compile-dictionary!
  "Compiles text translations stored in simple development-friendly Clojure map
  into form required by localized text translator.

    {:en {:example {:foo.html \"<tag>\"
                    :foo.note \"Translator note\"
                    :bar.md   \"**strong**\"
                    :baz      \"<tag>\"}}}
    =>
    {:en {:example/foo \"<tag>\"
          :example/bar \"<strong>strong</strong>\"
          :example/baz \"&lt;tag&gt;\"}}

  Note the optional key decorators."
  []
  (let [{:keys [dictionary dictionary-compiler-options]} @config]
    (->> (utils/leaf-paths dictionary)
         (map (partial compile-map-path dictionary-compiler-options))
         (apply merge-with merge) ; 1-level recursive merge
         (reset! compiled-dictionary))))

(compile-dictionary!) ; Actually compile default dictionary now

;; Automatically re-compile any time dictionary changes
(add-watch
 config "dictionary-watch"
 (fn [key ref old-state new-state]
   (when (or (not= (:dictionary old-state)
                   (:dictionary new-state))
             (not= (:dictionary-compiler-options old-state)
                   (:dictionary-compiler-options new-state)))
     (compile-dictionary!))))

(defn dictionary->xliff [m]) ; TODO Use hiccup?
(defn xliff->dictionary [s]) ; TODO Use clojure.xml/parse?

;;;; Translation

(defn- fqname
  "Like `name` but nil-safe and returns fully-qualified name."
  [keyword]
  (when keyword
    (if-let [ns (namespace keyword)]
      (str ns "/" (name keyword))
      (name keyword))))

(comment (fqname :a/b/c/d))

(def ^:private locales-to-check
  "Given a Locale, returns vector of dictionary locale names to check, in order
  of preference:
    #<Locale en_US_var1> => [:en_US_var1 :en_US :en]
    #<Locale en_US>      => [:en_US :en]
    #<Locale en>         => [:en]"
  (memoize
   (fn [Locale]
     (let [parts (str/split (str Locale) #"_")]
       (vec (for [n (range (count parts) 0 -1)]
              (keyword (str/join "_" (take n parts)))))))))

(comment (locales-to-check (parse-Locale :en_US)))

(defn t ; translate
  "Localized text translator. Takes a namespaced key :nsA/.../nsN within a scope
  :ns1/.../nsM and returns the best dictionary translation available for working
  locale.

  With additional arguments, treats translated text as pattern for message
  formatting."
  ([scoped-dict-key & args] (apply format-msg (t scoped-dict-key) args))
  ([scoped-dict-key]
     (let [fully-scoped-key ; :ns1/.../nsM/nsA/.../nsN = :ns1/.../nsN
           (if *translation-scope*
             (keyword (str (fqname *translation-scope*) "/"
                           (fqname scoped-dict-key)))
             scoped-dict-key)

           [lchoice1 lchoice2 lchoice3] (locales-to-check *Locale*)
           cdict-snap @compiled-dictionary]

       (or (get-in cdict-snap [lchoice1 fully-scoped-key])
           (when lchoice2 (get-in cdict-snap [lchoice2 fully-scoped-key]))
           (when lchoice3 (get-in cdict-snap [lchoice3 fully-scoped-key]))
           ((:missing-translation-fn @config) {:key     fully-scoped-key
                                               :locale *Locale*})))))

(comment (with-locale :en_ZA (t :example/foo))
         (with-locale :en_ZA (with-scope :example (t :foo)))
         (with-locale :en_ZA (t :invalid)))
